### PR TITLE
Adding AllBitsSet to IBinaryNumber Interface

### LINKED ISSUE #69676: [API Proposal]: Add AllBitsSet to the IBinaryNumber Interface
### Background and motivation

Conceptually, `T.AllBitsSet` returns an instance of the binary number type `T` where every bit is set to 1. For example, `Int32.AllBitsSet` is `0xFFFFFFFF`, or `-1`. This concept is generally useful for masking purposes. For example, Vectors use `AllBitsSet` often for Conditional Select, as bitwise-anding a number with `AllBitsSet` will produce the same number, in contrast to bitwise-anding a number with `Zero`. As brought up by https://github.com/dotnet/runtime/issues/60882, `AllBitsSet` would be a nice addition to the `IBinaryNumber` interface, potentially enabling future generic refactoring of `Scalar<T>`.

The following types would explicitly implement `AllBitsSet`: `Double`, `Half`, `NFloat`, `Single`, `Byte`, `Char`, `Int16`, `Int32`, `Int64`, `IntPtr`, `SByte`, `UInt16`, `UInt32`, `UInt64`, `UIntPtr`, `BigInteger`, `Int128`, `Uint128`.

Design of the implementation for all of these types is trivial except for `BigInteger`. For `BigInteger`, the most obvious implementation would be to return `MinusOne`, as despite the internal bit representation not being literally `AllBitsSet` (due to `BigInteger` having unbounded size), it behaves as such for bitwise operations.

### API Proposal

```csharp
namespace System.Numerics
{

  public partial interface IBinaryNumber<TSelf>
  {
       static abstract TSelf AllBitsSet { get; }
  }

}
```

### API Usage

```csharp

T genericAlgorithm(T input) 
where T : IBinaryNumber<T> {
    return input + T.AllBitsSet;
}
```

### Alternative Designs

When deciding whether a method is implemented explicitly or implicitly for Generic Math, we have used the following rule: if the concept is trivially handled by knowing the concrete type, it's implemented explicitly. In other words, it is not exposed on the concrete type. This way, the implementation is only available when working with the generic interface. 

In this case, we follow this logic and implement `AllBitsSet` explicitly, as knowing the concrete type leads to a trivial implementation (https://source.dot.net/#System.Private.CoreLib/Scalar.cs,11). However, it could be argued that `AllBitsSet` is conceptually tricky enough for some of these types that a implicit implementation is warranted.

### Risks

Currently, refactoring `Scalar<T>` to take advantage of this is not possible without a language change. More specifically, the language doesn't have a way to "bridge" type constraints. So, since `Vector<T>` is `where T : struct` but we'd want `Scalar<T>` to be `where T : struct, IBinaryNumber<T>`, we'd need the language to provide some way to allow `Vector<T>` to use `Scalar<T>` despite the mismatching type requirements.

### PR DESCRIPTION
**Pull request by dakersnar**
Fixes https://github.com/dotnet/runtime/issues/69676

Note: there is a trimmer bug that is preventing a DIM of AllBitsSet from being possible. This will be added later once the [bug](https://github.com/dotnet/linker/issues/2865) is fixed.

I tried to match the style of the files I was adding to, but feel free to nitpick any style choices. Specifically, there are quite a few ways to hard code in an AllBitsSet constant for each of these types, so happy to tweak as needed.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Double.cs [hunk 1]
```diff
@@ -547,6 +547,9 @@ object IConvertible.ToType(Type type, IFormatProvider? provider)
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static double IBinaryNumber<double>.AllBitsSet => BitConverter.UInt64BitsToDouble(0xFFFFFFFFFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Double.cs:551]: Adding the spacers makes it slightly more readable. I personally tend to put them in groups of `4`
```suggestion
        static double IBinaryNumber<double>.AllBitsSet => BitConverter.UInt64BitsToDouble(0xFFFF_FFFF_FFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/Int128.cs [hunk 2]
```diff
@@ -867,6 +867,9 @@ bool IBinaryInteger<Int128>.TryWriteLittleEndian(Span<byte> destination, out int
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static Int128 IBinaryNumber<Int128>.AllBitsSet => new Int128(0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Int128.cs:871]:
```suggestion
        static Int128 IBinaryNumber<Int128>.AllBitsSet => new Int128(0xFFFF_FFFF_FFFF_FFFF, 0xFFFF_FFFF_FFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs [hunk 3]
```diff
@@ -871,6 +871,18 @@ public int CompareTo(object? obj)
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static NFloat IBinaryNumber<NFloat>.AllBitsSet
+        {
+#if TARGET_64BIT
+            [NonVersionable]
+            get => (NFloat)BitConverter.UInt64BitsToDouble(0xFFFFFFFFFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:879]:
```suggestion
            get => (NFloat)BitConverter.UInt64BitsToDouble(0xFFFF_FFFF_FFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs [hunk 4]
```diff
@@ -871,6 +871,18 @@ public int CompareTo(object? obj)
...
+#else
+            [NonVersionable]
+            get => BitConverter.UInt32BitsToSingle(0xFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/NFloat.cs:882]:
```suggestion
            get => BitConverter.UInt32BitsToSingle(0xFFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/Single.cs [hunk 5]
```diff
@@ -539,6 +539,9 @@ object IConvertible.ToType(Type type, IFormatProvider? provider)
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static float IBinaryNumber<float>.AllBitsSet => BitConverter.UInt32BitsToSingle(0xFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Single.cs:543]:
```suggestion
        static float IBinaryNumber<float>.AllBitsSet => BitConverter.UInt32BitsToSingle(0xFFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/UInt128.cs [hunk 6]
```diff
@@ -906,6 +906,9 @@ bool IBinaryInteger<UInt128>.TryWriteLittleEndian(Span<byte> destination, out in
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static UInt128 IBinaryNumber<UInt128>.AllBitsSet => new UInt128(0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/UInt128.cs:910]:
```suggestion
        static UInt128 IBinaryNumber<UInt128>.AllBitsSet => new UInt128(0xFFFF_FFFF_FFFF_FFFF, 0xFFFF_FFFF_FFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/UInt64.cs [hunk 7]
```diff
@@ -358,6 +358,9 @@ bool IBinaryInteger<ulong>.TryWriteLittleEndian(Span<byte> destination, out int
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static ulong IBinaryNumber<ulong>.AllBitsSet => 0xFFFFFFFFFFFFFFFF;
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/UInt64.cs:362]:
```suggestion
        static ulong IBinaryNumber<ulong>.AllBitsSet => 0xFFFF_FFFF_FFFF_FFFF;
```

or same suggestion as above

File: src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs [hunk 8]
```diff
@@ -344,6 +344,18 @@ bool IBinaryInteger<nuint>.TryWriteLittleEndian(Span<byte> destination, out int
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static nuint IBinaryNumber<nuint>.AllBitsSet
+        {
+#if TARGET_64BIT
+            [NonVersionable]
+            get => unchecked((nuint)0xFFFFFFFFFFFFFFFF);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs:352]:
```suggestion
            get => unchecked((nuint)0xFFFF_FFFF_FFFF_FFFF);
```

File: src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs [hunk 9]
```diff
@@ -344,6 +344,18 @@ bool IBinaryInteger<nuint>.TryWriteLittleEndian(Span<byte> destination, out int
...
+#else
+            [NonVersionable]
+            get => (nuint)0xFFFFFFFF;
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/UIntPtr.cs:355]:
```suggestion
            get => (nuint)0xFFFF_FFFF;
```

File: src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs [hunk 10]
```diff
@@ -217,6 +217,22 @@ public static void AdditiveIdentityTest()
...
+            {
+                Assert.Equal(0xFFFFFFFF, BitConverter.SingleToUInt32Bits((float)BinaryNumberHelper<NFloat>.AllBitsSet));
+                Assert.Equal((uint)0, (uint)(BitConverter.SingleToUInt32Bits((float)BinaryNumberHelper<NFloat>.AllBitsSet) + 1));
+            }
+            else
+            {
+                Assert.Equal(0xFFFFFFFFFFFFFFFF, BitConverter.DoubleToUInt64Bits((double)BinaryNumberHelper<NFloat>.AllBitsSet));
+                Assert.Equal((ulong)0, (ulong)(BitConverter.DoubleToUInt64Bits((double)BinaryNumberHelper<NFloat>.AllBitsSet) + 1));
+            }
+
```
Comment (tannergooding) [inline src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs:233]: nit: For consistency with the other tests, do:
```suggestion
            if (Environment.Is64BitProcess)
            {
                Assert.Equal(0xFFFFFFFFFFFFFFFF, BitConverter.DoubleToUInt64Bits((double)BinaryNumberHelper<NFloat>.AllBitsSet));
                Assert.Equal((ulong)0, (ulong)(BitConverter.DoubleToUInt64Bits((double)BinaryNumberHelper<NFloat>.AllBitsSet) + 1));
            }
            else
            {
                Assert.Equal(0xFFFFFFFF, BitConverter.SingleToUInt32Bits((float)BinaryNumberHelper<NFloat>.AllBitsSet));
                Assert.Equal((uint)0, (uint)(BitConverter.SingleToUInt32Bits((float)BinaryNumberHelper<NFloat>.AllBitsSet) + 1));
            }
```

Comment (tannergooding) [reply src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs:233]: I'm also not convinced the assert that `AllBitsSet + 1 == 0` is adding much value.

It's not harming anything either however. You could simplify it a bit by using `0UL`, `0U`, `1UL`, and `1U` however.

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs (hunk 10 above)
Comment (dakersnar) [reply src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs:233]: Thanks for the suggestions, I'll add them all. 

Re: the tests, I wanted to include a test case that wasn't a copy paste of my implementation logic. For example, if I implement `char.AllBitsSet` as `0xFF`, then make a test case that only does `Assert.Equal(0xFF, char.AllBitsSet)`, the test case passes, and the bug goes unnoticed. With `AllBitsSet + 1 == 0`, I can confirm that my implementation logic is actually correct. This example was actually a real bug I ran into, as I was going off of my memory of C and assuming `char` was one byte.

Comment (tannergooding) [reply src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs:233]: Makes sense, but I'd note that testing `Zero == ~AllBitsSet` might be a better one.

There are theoretically types where `AllBitsSet + 1` won't become the same as "zero", for many possible reasons. However, for bitwise manipulation the implication is that `Zero` is "no bits set" and `~` being a one's complement should therefore give you "all bits set" as the result (regardless of underlying format representation), regardless of there being other unrelated bits (like flags, etc).

File: src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs [hunk 11]
```diff
@@ -10,6 +10,9 @@ public interface IBinaryNumber<TSelf>
           INumber<TSelf>
         where TSelf : IBinaryNumber<TSelf>
     {
+        /// <summary>Gets an instance of the binary type in which all bits are set.</summary>
+        static abstract TSelf AllBitsSet { get; } // TODO: add the DIM once https://github.com/dotnet/linker/issues/2865 is fixed
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: Do we have a runtime issue tracking all of the APIs that need to be made DIMs once this bug is fixed?  Presumably this is a must-do for .NET 7?

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: The other APIs are already DIM where relevant. This is a new scenario specific to properties.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: > The other APIs are already DIM where relevant.

Oh.  We're not making APIs like IParseable.Parse DIMs?
https://github.com/dotnet/runtime/blob/3fb9511bf78a5de5cb73bad8fd50a69c691d61cd/src/libraries/System.Private.CoreLib/src/System/IParsable.cs#L20

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: (or agree that we just always throw `FormatException` and expect users that need to differentiate to always override).

Comment (dakersnar) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/IBinaryNumber.cs:14]: The fact that I'm the first person to run into this implies this is the first static virtual property somebody has tried to use. This is surprising and suspicious. I wonder if I'm just the first one to realize the DIM isn't being overridden by inheriting classes. Regardless, yes, I think fixing the bug should have relatively high priority.

File: src/libraries/System.Private.CoreLib/src/System/UInt16.cs [hunk 12]
```diff
@@ -373,6 +373,9 @@ bool IBinaryInteger<ushort>.TryWriteLittleEndian(Span<byte> destination, out int
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static ushort IBinaryNumber<ushort>.AllBitsSet => 0xFFFF;
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/UInt16.cs:377]:
```suggestion
        static ushort IBinaryNumber<ushort>.AllBitsSet => ushort.MaxValue;
```

File: src/libraries/System.Private.CoreLib/src/System/UInt32.cs [hunk 13]
```diff
@@ -359,6 +359,9 @@ bool IBinaryInteger<uint>.TryWriteLittleEndian(Span<byte> destination, out int b
         // IBinaryNumber
         //
 
+        /// <inheritdoc cref="IBinaryNumber{TSelf}.AllBitsSet" />
+        static uint IBinaryNumber<uint>.AllBitsSet => 0xFFFFFFFF;
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/UInt32.cs:363]:
```suggestion
        static uint IBinaryNumber<uint>.AllBitsSet => uint.MaxValue;
```

File: src/libraries/System.Private.CoreLib/src/System/UInt64.cs (hunk 7 above)
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/UInt64.cs:362]:
```suggestion
        static ulong IBinaryNumber<ulong>.AllBitsSet => ulong.MaxValue;
```

File: src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs [hunk 14]
```diff
@@ -217,6 +217,22 @@ public static void AdditiveIdentityTest()
         // IBinaryNumber
         //
 
+        [Fact]
+        public static void AllBitsSetTest()
+        {
+            if (NFloat.Size == sizeof(float))
+            {
+                Assert.Equal(0xFFFFFFFF, BitConverter.SingleToUInt32Bits((float)BinaryNumberHelper<NFloat>.AllBitsSet));
```
Comment (dakersnar) [inline src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs:225]:
```suggestion
                Assert.Equal(0xFFFF_FFFF, BitConverter.SingleToUInt32Bits((float)BinaryNumberHelper<NFloat>.AllBitsSet));
```

File: src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs (hunk 10 above)
Comment (dakersnar) [reply src/libraries/System.Runtime.InteropServices/tests/System.Runtime.InteropServices.UnitTests/System/Runtime/InteropServices/NFloatTests.GenericMath.cs:233]: Good point, I'll switch to that method

**Review by dakersnar — COMMENTED** [review]
More suggestions from code review

**Review by tannergooding — APPROVED** [review]

**Pull request merged**